File: backend/app/schemas/workflow.py
from datetime import datetime
from typing import Optional, List, Dict, Any

from pydantic import BaseModel, Field


# ── Workflow Node Schemas ──────────────────────────────────────────────────────

class WorkflowNodeBase(BaseModel):
    node_type: str = Field(..., description="节点类型")
    name: str = Field(..., description="节点名称")
    position_x: float = Field(default=0.0, description="X坐标")
    position_y: float = Field(default=0.0, description="Y坐标")
    config: Optional[Dict[str, Any]] = Field(default=None, description="节点配置")


class WorkflowNodeCreate(WorkflowNodeBase):
    pass


class WorkflowNodeUpdate(WorkflowNodeBase):
    node_type: Optional[str] = Field(default=None, description="节点类型")
    name: Optional[str] = Field(default=None, description="节点名称")


class WorkflowNodeResponse(WorkflowNodeBase):
    id: int
    workflow_id: int
    tenant_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ── Workflow Edge Schemas ──────────────────────────────────────────────────────

class WorkflowEdgeBase(BaseModel):
    source_node_id: int = Field(..., description="源节点ID")
    target_node_id: int = Field(..., description="目标节点ID")
    condition: Optional[Dict[str, Any]] = Field(default=None, description="条件配置")
    label: Optional[str] = Field(default=None, description="边标签")


class WorkflowEdgeCreate(WorkflowEdgeBase):
    pass


class WorkflowEdgeUpdate(WorkflowEdgeBase):
    source_node_id: Optional[int] = Field(default=None, description="源节点ID")
    target_node_id: Optional[int] = Field(default=None, description="目标节点ID")


class WorkflowEdgeResponse(WorkflowEdgeBase):
    id: int
    workflow_id: int
    tenant_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# ── Workflow Schemas ────────────────────────────────────────────────────────────

class WorkflowBase(BaseModel):
    name: str = Field(..., description="工作流名称")
    description: Optional[str] = Field(default=None, description="工作流描述")


class WorkflowCreate(WorkflowBase):
    nodes: Optional[List[WorkflowNodeCreate]] = Field(default=None, description="节点列表")
    edges: Optional[List[WorkflowEdgeCreate]] = Field(default=None, description="边列表")


class WorkflowUpdate(WorkflowBase):
    name: Optional[str] = Field(default=None, description="工作流名称")
    description: Optional[str] = Field(default=None, description="工作流描述")
    status: Optional[str] = Field(default=None, description="工作流状态")
    nodes: Optional[List[WorkflowNodeCreate]] = Field(default=None, description="节点列表")
    edges: Optional[List[WorkflowEdgeCreate]] = Field(default=None, description="边列表")


class WorkflowResponse(WorkflowBase):
    id: int
    tenant_id: int
    status: str
    is_active: bool
    created_by: Optional[int]
    created_at: datetime
    updated_at: datetime
    deleted_at: Optional[datetime]
    nodes: List[WorkflowNodeResponse] = []
    edges: List[WorkflowEdgeResponse] = []

    class Config:
        from_attributes = True


class WorkflowListResponse(BaseModel):
    items: List[WorkflowResponse]
    total: int
    page: int
    page_size: int


# ── Workflow Execution Schemas ──────────────────────────────────────────────────

class WorkflowExecutionRequest(BaseModel):
    input_data: Optional[Dict[str, Any]] = Field(default=None, description="输入参数")


class WorkflowExecutionResponse(BaseModel):
    id: int
    workflow_id: int
    tenant_id: int
    status: str
    input_data: Optional[Dict[str, Any]]
    output_data: Optional[Dict[str, Any]]
    error_message: Optional[str]
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    created_by: Optional[int]
    created_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/types/workflow.ts
// ── Workflow Node Types ───────────────────────────────────────────────────────

export type NodeType = 'start' | 'end' | 'llm' | 'condition' | 'knowledge' | 'code' | 'tool' | 'loop' | 'variable'

export interface WorkflowNodeConfig {
  // LLM节点配置
  model_id?: string
  prompt_template?: string
  temperature?: number
  max_tokens?: number
  output_variable?: string

  // Condition节点配置
  conditions?: Array<{
    expression: string
    label: string
  }>

  // Knowledge节点配置
  knowledge_base_id?: string
  query_template?: string
  top_k?: number

  // Code节点配置
  code?: string
  input_variables?: string[]
  code_output_variable?: string

  // Tool节点配置
  tool_type?: string
  tool_name?: string
  tool_config?: Record<string, unknown>

  // Loop节点配置
  loop_type?: 'for' | 'while'
  loop_variable?: string
  loop_source?: string
  max_iterations?: number

  // Variable节点配置
  variables?: Array<{
    name: string
    type: 'static' | 'context' | 'expression'
    value?: unknown
    source?: string
    expression?: string
  }>

  // Start/End节点配置
  output_variables?: Array<{
    name: string
    source: string
    value?: unknown
  }>
}

export interface WorkflowNode {
  id: string
  workflow_id: string
  tenant_id: string
  node_type: NodeType
  name: string
  position_x: number
  position_y: number
  config: WorkflowNodeConfig | null
  created_at: string
  updated_at: string
}

export interface WorkflowNodeCreate {
  node_type: NodeType
  name: string
  position_x: number
  position_y: number
  config?: WorkflowNodeConfig
  temp_id?: string // 临时节点ID（用于新建节点时的ID映射）
}

// ── Workflow Edge Types ───────────────────────────────────────────────────────

export interface WorkflowEdgeCondition {
  expression: string
  label: string
}

export interface WorkflowEdge {
  id: string
  workflow_id: string
  tenant_id: string
  source_node_id: string
  target_node_id: string
  condition: WorkflowEdgeCondition | null
  label: string | null
  created_at: string
  updated_at: string
}

export interface WorkflowEdgeCreate {
  source_node_id: string | number // 可以是临时ID字符串或真实ID数字
  target_node_id: string | number // 可以是临时ID字符串或真实ID数字
  condition?: WorkflowEdgeCondition
  label?: string
}

// ── Workflow Types ─────────────────────────────────────────────────────────────

export type WorkflowStatus = 'draft' | 'published' | 'archived'

export interface Workflow {
  id: string
  tenant_id: string
  name: string
  description: string | null
  status: WorkflowStatus
  is_active: boolean
  created_by: number | null
  created_at: string
  updated_at: string
  deleted_at: string | null
  nodes: WorkflowNode[]
  edges: WorkflowEdge[]
}

export interface WorkflowCreateRequest {
  name: string
  description?: string
  nodes?: WorkflowNodeCreate[]
  edges?: WorkflowEdgeCreate[]
}

export interface WorkflowUpdateRequest {
  name?: string
  description?: string
  status?: WorkflowStatus
  nodes?: WorkflowNodeCreate[]
  edges?: WorkflowEdgeCreate[]
}

export interface WorkflowListResponse {
  items: Workflow[]
  total: number
  page: number
  page_size: number
}

// ── Workflow Execution Types ───────────────────────────────────────────────────

export type ExecutionStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'

export interface WorkflowExecution {
  id: string
  workflow_id: string
  tenant_id: string
  status: ExecutionStatus
  input_data: Record<string, unknown> | null
  output_data: Record<string, unknown> | null
  error_message: string | null
  started_at: string | null
  completed_at: string | null
  created_by: number | null
  created_at: string
}

export interface WorkflowExecutionRequest {
  input_data?: Record<string, unknown>
}

// ── SSE Event Types ───────────────────────────────────────────────────────────

export interface SSEExecutionStartedEvent {
  type: 'execution_started'
  execution_id: string
  workflow_id: string
}

export interface SSENodeStartedEvent {
  type: 'node_started'
  node_id: string
  node_name: string
  node_type: NodeType
}

export interface SSENodeCompletedEvent {
  type: 'node_completed'
  node_id: string
  node_name: string
  output: Record<string, unknown>
}

export interface SSENodeFailedEvent {
  type: 'node_failed'
  node_id: string
  node_name: string
  error: string
}

export interface SSEExecutionCompletedEvent {
  type: 'execution_completed'
  execution_id: string
  output: Record<string, unknown>
}

export interface SSEExecutionFailedEvent {
  type: 'execution_failed'
  execution_id: string
  error: string
}

export type SSEWorkflowEvent =
  | SSEExecutionStartedEvent
  | SSENodeStartedEvent
  | SSENodeCompletedEvent
  | SSENodeFailedEvent
  | SSEExecutionCompletedEvent
  | SSEExecutionFailedEvent